### ISSUE TITLE
There should be a handler which is guaranteed to be called when the request is done

### ISSUE THREAD

**Issue by luengnat** [eclipse-vertx/vert.x#2524]
There should be a handler which is guaranteed to be called when the request is done so that we can perform the resource cleanup. Currently, there are some hooks but it is hard to deal with. 

On the server side, you have .addBodyEndHandler() and .closeHandler(). On the client side, it is even more confusing since you have to hook both to the HttpClientRequest.exceptionHandler() and HttpClientResponse.exceptionHandler(), HttpClientResponse.endHandler() while it does not even have an analogous .closeHandler().

**Comment by vietj** [comment]
I agree we should need more reliability and consistency and that's something we should address at least for Vert.x 4.

For Vert.x 4 we will remove the `HttpClientRequest#endHandler` because it's very confusing and the API will be changed to use an `Handler<AsyncResult>` method.

`HttpClientResponse#endHandler` is actually defined by `ReadStream` and the end handler is not guaranteed to be called always, which is why one has to use currently exception handler and end handler.

**Comment by vietj** [comment]
I think the new response Future provides a reliable way to implement this
